polkadot/primitives/src/vstaging/mod.rs

```diff
@@ -98,3 +116,873 @@ impl<BlockNumber: Default + From<u32>> Default for SchedulerParams<BlockNumber>
+		match value {
+			CandidateEvent::CandidateBacked(receipt, head_data, core_index, group_index) =>
+				super::v7::CandidateEvent::CandidateBacked(
+					receipt.into(),
+					head_data,
+					core_index,
+					group_index,
+				),
+			CandidateEvent::CandidateIncluded(receipt, head_data, core_index, group_index) =>
+				super::v7::CandidateEvent::CandidateIncluded(
+					receipt.into(),
+					head_data,
+					core_index,
+					group_index,
+				),
+			CandidateEvent::CandidateTimedOut(receipt, head_data, core_index) =>
+				super::v7::CandidateEvent::CandidateTimedOut(receipt.into(), head_data, core_index),
+		}
+	}
+}
+
+impl<H> CandidateReceiptV2<H> {
+	/// Get a reference to the candidate descriptor.
+	pub fn descriptor(&self) -> &CandidateDescriptorV2<H> {
+		&self.descriptor
+	}
+
+	/// Computes the blake2-256 hash of the receipt.
+	pub fn hash(&self) -> CandidateHash
+	where
+		H: Encode,
+	{
+		CandidateHash(BlakeTwo256::hash_of(self))
+	}
+}
+
+impl<H: Clone> CommittedCandidateReceiptV2<H> {
+	/// Transforms this into a plain `CandidateReceipt`.
+	pub fn to_plain(&self) -> CandidateReceiptV2<H> {
+		CandidateReceiptV2 {
+			descriptor: self.descriptor.clone(),
+			commitments_hash: self.commitments.hash(),
+		}
+	}
+
+	/// Computes the hash of the committed candidate receipt.
+	///
+	/// This computes the canonical hash, not the hash of the directly encoded data.
+	/// Thus this is a shortcut for `candidate.to_plain().hash()`.
+	pub fn hash(&self) -> CandidateHash
+	where
+		H: Encode,
+	{
+		self.to_plain().hash()
+	}
+
+	/// Does this committed candidate receipt corresponds to the given [`CandidateReceipt`]?
+	pub fn corresponds_to(&self, receipt: &CandidateReceiptV2<H>) -> bool
+	where
+		H: PartialEq,
+	{
+		receipt.descriptor == self.descriptor && receipt.commitments_hash == self.commitments.hash()
+	}
+}
+
+impl PartialOrd for CommittedCandidateReceiptV2 {
+	fn partial_cmp(&self, other: &Self) -> Option<core::cmp::Ordering> {
+		Some(self.cmp(other))
+	}
+}
+
+impl Ord for CommittedCandidateReceiptV2 {
+	fn cmp(&self, other: &Self) -> core::cmp::Ordering {
+		self.descriptor
+			.para_id
+			.cmp(&other.descriptor.para_id)
+			.then_with(|| self.commitments.head_data.cmp(&other.commitments.head_data))
+	}
+}
+
+impl<H: Copy> From<CommittedCandidateReceiptV2<H>> for super::v7::CommittedCandidateReceipt<H> {
+	fn from(value: CommittedCandidateReceiptV2<H>) -> Self {
+		Self { descriptor: value.descriptor.into(), commitments: value.commitments }
+	}
+}
+
+impl<H: Copy> From<CandidateReceiptV2<H>> for super::v7::CandidateReceipt<H> {
+	fn from(value: CandidateReceiptV2<H>) -> Self {
+		Self { descriptor: value.descriptor.into(), commitments_hash: value.commitments_hash }
+	}
+}
+
+/// A strictly increasing sequence number, tipically this would be the least significant byte of the
+/// block number.
+#[derive(PartialEq, Eq, Clone, Encode, Decode, TypeInfo, RuntimeDebug)]
+pub struct CoreSelector(pub u8);
```

Comment:
Didn't we decide to use u16? AFAICT this should be at most the highest number of cores a para can have assigned.
the reality is nowhere near 256 now but u16 would be more future-proof

Reply:
I think `u8` is better because it saves 1 byte that is stored on chain. Parachains needing more than 256 cores is a nice problem to have. If that happens in the future we should already have a more flexible and future proof format for the descriptor and we should have already implemented PVF input/output versioning. There is also the possibility to introduce a `CoreSelectorExt` message to solve it in the future.

Reply:
fair enough